src/test/functions/env.spec.ts

```diff
@@ -0,0 +1,164 @@
+import { expect } from "chai";
+import * as sinon from "sinon";
+
+import { FirebaseError } from "../../error";
+import * as fenv from "../../functions/env";
+import * as envstore from "../../functions/envstore";
+
+describe("function env", () => {
+  describe("parseKvArgs", () => {
+    it("should successfully parse kv args", () => {
+      const args = ["FOO=bar", "BAR=foo=bar"];
+      expect(fenv.parseKvArgs(args)).to.deep.equal({ FOO: "bar", BAR: "foo=bar" });
+    });
+
+    it("should throw error given invalid keys", () => {
+      const args = ["FOO=bar", "X_GOOGLE_BAR=foo=bar"];
+      expect(() => {
+        fenv.parseKvArgs(args);
+      }).to.throw(FirebaseError);
+    });
+  });
+
+  describe("validateKey", () => {
+    it("should accept valid keys", () => {
+      const keys = ["FOO", "ABC_EFG", "A1_B2"];
+      keys.forEach((key) => {
+        expect(() => {
+          fenv.validateKey(key);
+        }).not.to.throw;
+      });
+    });
+
+    it("should throw error given invalid keys", () => {
+      const keys = ["", "A", "1F", "B=C"];
+      keys.forEach((key) => {
+        expect(() => {
+          fenv.validateKey(key);
+        }).not.to.throw;
+      });
+    });
+
+    it("should throw error given reserved keys", () => {
+      const keys = [
+        "FIREBASE_CONFIG",
+        "FUNCTION_TARGET",
+        "FUNCTION_SIGNATURE_TYPE",
+        "K_SERVICE",
+        "K_REVISION",
+        "PORT",
+        "K_CONFIGURATION",
+      ];
+      keys.forEach((key) => {
+        expect(() => {
+          fenv.validateKey(key);
+        }).to.throw(FirebaseError);
+      });
+    });
+
+    it("should throw error given keys with reserved prefix", () => {
+      expect(() => {
+        fenv.validateKey("X_GOOGLE_");
+      }).to.throw(FirebaseError);
+
+      expect(() => {
+        fenv.validateKey("X_GOOGLE_FOOBAR");
+      }).to.throw(FirebaseError);
+    });
+  });
+
+  describe("formatEnv", () => {
+    it("should format env object in a console-friendly way", () => {
+      expect(fenv.formatEnv({ FOO: "bar", AB1: "EFG" })).to.equal("FOO=bar\nAB1=EFG");
+    });
+  });
+
+  describe("cloneEnvs", () => {
+    let createStore: sinon.SinonStub;
+    let deleteStore: sinon.SinonStub;
+    let getStore: sinon.SinonStub;
+
+    beforeEach(() => {
+      createStore = sinon.stub(envstore, "createStore").rejects("Unexpected call");
+      deleteStore = sinon.stub(envstore, "deleteStore").rejects("Unexpected call");
+      getStore = sinon.stub(envstore, "getStore").rejects("Unexpected call");
+    });
+
+    afterEach(() => {
+      createStore.restore();
+      deleteStore.restore();
+      getStore.restore();
+    });
+
+    it("should clone all environment variables from the source project", async () => {
+      const envs = {
+        FOO: "bar",
+        AB1: "ab1",
+      };
+      const fromP = "project-1";
+      const toP = "project-2";
+
+      createStore.onFirstCall().resolves({ vars: envs });
```

Comment:
Getting very nitty here, but:

1. It might be clearer if these API calls are in sequence (delete before create)
2. It might be clearer to replace `onFirstCall` with `withArgs`
3. Since the order of these API calls matters (e.g. delete before create) you might want to add
`expect(deleteStore.calledBefore(createStore))`